Updated the Google Analytics tracking for a non-interaction hit

kingo55 opened the pull request:
Hey James,

Excellent script you have here. I'm going to make a few changes, this is the first. The event tracking used to affect the bounce rate metric. Now it does not.

Have also started a blog post about this. I bet my readers will love this.

Rob

kingo55 committed:
Updated event tracking to set non-interaction hit

The event tracking used to affect the bounce rate metric. Now it does not.

jamesyu commented on lib/cohorts.js:
```diff
@@ -5,22 +5,22 @@ Cohorts = (function() {
         onInitialize: function(inTest, testName, cohort) {
             if(inTest) {
-                this.trackEvent(this.nameSpace, testName, cohort + ' | Total');
+                this.trackEvent(this.nameSpace, testName, cohort + ' | Total',,true);
```
you have double commas here

kingo55 replied:
This is because we're not passing a value to opt_value for the track event function. We need to send "true" to int_hit, though. See opt_noninteraction here: https://developers.google.com/analytics/devguides/collection/gajs/eventTrackerGuide

What do you suggest here, James?

jamesyu commented on lib/cohorts.js:
```diff
@@ -5,22 +5,22 @@ Cohorts = (function() {
         },
         onEvent: function(testName, cohort, eventName) {
-            this.trackEvent(this.nameSpace, testName, cohort + ' | ' + eventName);
+            this.trackEvent(this.nameSpace, testName, cohort + ' | ' + eventName,,false);
```
same here

kingo55 committed:
Added custom variable tracking

Under storage adapter, I have added the custom variable slot and setCustomVar function.

Under the test object, I have created a new variable called scope. I intend to make this work with a cookie function. i.e. depending on the scope (1: visitor, 2: session, 3: page), the cookies will persist for that particular scope.

kingo55 committed:
Updated Readme with changes to the test code

Have included the following new features:

* Custom variable tracking
* Setting the test's scope (e.g. persist cookies for page, session or visitor lifetime)
* Removed a bunch of code not in use
* Made it mandatory to include a storage adapter with each experiment

kingo55 committed:
Beautified code to make it easier to read

kingo55 committed:
Updating test.html

kingo55 committed:
Updating cohorts.js

Have included the following in this update:

* Custom variable tracking
* Setting the test's scope (e.g. persist cookies for page, session or visitor lifetime)
* Removed a bunch of code not in use
* Made it mandatory to include a storage adapter with each experiment

jamesyu commented on README.textile:
```diff
@@ -19,44 +19,65 @@ h2. Basic Usage
+            },
+        },
+
+        storageAdapter: {
```
is there a reason you aren't just using the google analytics adapter?

jamesyu commented on README.textile:
```diff
@@ -19,44 +19,65 @@ h2. Basic Usage
-    $('#small').click(function() {
-        header_test.event('Clicked on Header');
+    $('p.message').click(function () {
+        hello - goodbye.event('Clicked on message'); // Track any evens with your storage adapter
```
this doesn't look right.. hello - goodbye.event?

jamesyu commented on README.textile:
```diff
@@ -84,23 +106,22 @@ Google Analytics event tracking is a great way to store data. They provide a gre
-    //   eventName: the name of the event the visitor triggered
-    onEvent: function(testName, cohort, eventName) {
-        // do event stuff
+    storageAdapter: {
```
this section should describe what the storage adapter looks like in general.. why are you showing a specific example here?

jamesyu commented:
there are a bunch of issues with this pull request - specifically, i would like to keep the google analytics adapter as is.

kingo55 commented:
Thanks for the feedback, James.

I'll see about fixing those issues up. In retrospect removing the storage
adapter wasn't the brightest move. I will, however need to define the
custom variable slot number (cv_slot) when creating the test object.

As far as the way I'm which the experiment scope is defined (persisting
over the duration of a visitor, session or page view), are you happy with
that? I figure this kills two birds with one stone. We can set the cookie
expiry and set the scope of the custom variable in one go.
On 11/03/2013 8:11 AM, "James Yu" notifications@github.com wrote:

> there are a bunch of issues with this pull request - specifically, i would
> like to keep the google analytics adapter as is.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jamesyu/cohorts/pull/2#issuecomment-14689361
> .

jamesyu commented:
Yep, I like the scoping, that's very useful.

kingo55 committed:
Added GoogleAnalyticsAdapter back in

Also defining cv_slot within the test object rather than the storage adapter object.

kingo55 committed:
Updated Readme with James' suggestions

kingo55 committed:
Updated test.html with changes to script

kingo55 committed:
Update storage adapter example

kingo55 committed:
Removed jQuery Library

No real need for this in here.

kingo55 commented:
Thanks again for the feedback, James. I've made some commits to address your suggestions and add a little more polish to my work. I'm really enjoying using the script and running experiments across a number of my sites.

Anyway, does the script look sufficient to you now?

_Blog post about cohorts.js_
Also of interest, you may like to see the blog post I wrote about the project (of course, when the main project includes the updates, I'll switch the the link back): 

http://www.optimisationbeacon.com/testing/using-cohorts-js-an-open-source-javascript-ab-mvt-library/

Of particular interest, I suggest a number of improvements which could add some additional value to the framework.

kingo55 committed:
Enabled preview mode when sample is set to 0

Preview mode should run irregardless of the sample size. This is what you would expect to see in a preview mode.

kingo55 committed:
Updating cohorts.js

Adding loads of new features to the test object:

- User agent match
- Cookie match
- Referrer match
- URL match (for A/B URL redirects)
- DOM Ready function to ensure code runs as soon as all elements exist in the DOM
- Also made some cosmetic changes to the GA code

kingo55 committed:
Delete

kingo55 committed:
Ensuring visitors excluded from the test are not allocated to the test

kingo55 committed:
Update to test logic, datalayer as storage adapter and better comments

Lots of small little tweaks here and there.

kingo55 committed:
Create cohorts.js

kingo55 committed:
Fixing user agent blocking

kingo55 committed:
Update test.html

kingo55 committed:
Fixing user agent blocking

kingo55 committed:
Update README.textile

cclafferty committed:
Fixes referrer being matched when passed an empty array

The referrer array is queried even when it's empty. Other filters ignore
empty arrays, so this fix is more for consistency.

kingo55 committed:
Merge pull request #1 from cclafferty/master

Fixes referrer being matched when passed an empty array

mihalik committed:
Fixing typo in UA detection for ios devices

kingo55 committed:
Merge pull request #2 from mihalik/master

Fixing typo in UA detection for ios devices

kingo55 committed:
Updating object code

Had an unnecessary comma in there after the User Agent blocking

allmywant committed:
support percentage based random cohort choosen

kingo55 committed:
Merge pull request #3 from allmywant/master

Fine tune the percentage of visitors allocated to each cohort. I.e. Control: 20%, Treatment: 80%

allmywant committed:
Add 'New to this test' feature

allmywant committed:
Merge remote-tracking branch 'upstream/master'

allmywant committed:
add excludeBySample feature

implement Visitor must not be allowed in the test now or in the future,
when Excluded by sample parameter.

kingo55 committed:
Merge pull request #4 from allmywant/master

Add 'New to this test' feature

kingo55 committed:
Update cohorts.js

Updating storage adapter to pass newToTest into datalayer

allmywant committed:
Cohorts improvements

1. rewirte Cookies object, small and faster
2. remove unused functions(inChort, Util.size)
3. other minor changes to improve the performence

kingo55 committed:
Merge pull request #7 from allmywant/master

Optimisations to the Cohorts library from David.

Thanks, @allmywant!

allmywant committed:
Redefining variable names in test object

Updated values
name -> "name"
scope -> "testScope":"user" // Values should include "user", "session"
and "hit". Default should be "user" if this variable is not set.
varSlot -> "analyticsSlot"
sample -> "sampleRate"
urlMatch -> "targetURL"
uaBlock -> "userAgentExclude"
cookieMatch -> "cookieMatch"
referrerMatch -> "referrerMatch"
cohorts -> "cohorts"

URL match object
regex -> "expression": /^\/store\/tyres\/.*/i,
regexType -> "match":"pathname|href"

kingo55 committed:
Merge pull request #11 from allmywant/master

Redefining variable names in test object

allmywant committed:
support same simpleRate values

kingo55 committed:
Merge pull request #14 from allmywant/master

support same simpleRate values
